src/Data/Constraint/Extras.hs

```diff
@@ -27,105 +28,73 @@
     -- * Bringing instances into scope
-  , Has
-  , has
   , Has'
   , has'
   , HasV
   , hasV
   , whichever
     -- * Misc
   , Implies1(..)
-    -- * Deprecated
-  , ArgDictV
   ) where
 
 import Data.Constraint
 import Data.Constraint.Compose
 import Data.Constraint.Flip
 import Data.Constraint.Forall
-import Data.Kind
 
--- | Morally, this class is for GADTs whose indices can be finitely
--- enumerated. An @'ArgDict' c f@ instance allows us to do two things:
---
--- 1. 'ConstraintsFor' requests the set of constraints @c a@ for all
---    possible types @a@ that could be applied to @f@.
---
--- 2. 'argDict' selects a specific @c a@ given a value of type @f a@.
---
--- Use 'Data.Constraint.Extras.TH.deriveArgDict' to derive instances
--- of this class.
-class ArgDict (c :: k -> Constraint) (f :: k -> Type) where
-  -- | Apply @c@ to each possible type @a@ that could appear in a @f a@.
+-- | The constraint @Has c f@ means that given any value of type @f a@, we can determine
+-- that there is an instance of @c a@. For example, @Has Show Tag@ means that given any
+-- @x :: Tag a@, we can conclude @Show a@. Most commonly, the type @f@ will be a GADT,
+-- where we can enumerate all the possible index types through pattern matching, and
+-- discover that there is an appropriate instance in each case. In this sort of
+-- situation, the @c@ can be left entirely polymorphic in the instance for @Has@, and
+-- this is the sort of instance that the provided Template Haskell code writes.
+
+-- It is also sometimes possible to hand-write instances of @Has c f@ for specific
+-- classes @c@ in cases where @f@ is a data type that packs an appropriate dictionary
+-- into its constructors.
+class Has c f where
+  -- | Use the @f a@ to show that there is an instance of @c a@, and
+  -- bring it into scope.
   --
-  -- > ConstraintsFor Show Tag = (Show Int, Show Bool)
-  type ConstraintsFor f c :: Constraint
-
-  -- | Use an @f a@ to select a specific dictionary from @ConstraintsFor f c@.
+  -- The order of type variables is chosen to work
+  -- with @-XTypeApplications@.
   --
-  -- > argDict I :: Dict (Show Int)
-  argDict :: ConstraintsFor f c => f a -> Dict (c a)
+  -- > -- Hold a value of type a, along with a tag identifying the a.
+  -- > data SomeTagged tag where
+  -- >   SomeTagged :: a -> tag a -> SomeTagged tag
+  -- >
+  -- > -- Use the stored tag to identify the thing we have, allowing us to call 'show'. Note that we
+  -- > -- have no knowledge of the tag type.
+  -- > showSomeTagged :: Has Show tag => SomeTagged tag -> String
+  -- > showSomeTagged (SomeTagged a tag) = has @Show tag $ show a
+  has :: forall a r. f a -> (c a => r) -> r
+  has x r | Dict <- argDict @c x = r
+
+  -- | Use an @f a@ to obtain a dictionary for @c a@
+  --
+  -- > argDict @Show I :: Dict (Show Int)
+  argDict :: forall a. f a -> Dict (c a)
+  argDict x = has @c x Dict
+  {-# MINIMAL has | argDict #-}
 
--- | \"Primed\" variants (@ConstraintsFor'@, 'argDict'', 'Has'',
```

Comment:
Although ConstraintsFor' goes away, it might be nice to say what "primed" variants and "V" variants of things mean.

Reply:
I do, just in the documentation for those things. But I can document this type synonym as well.